### ISSUE TITLE
Add support for reserved capacity

### ISSUE THREAD

**Issue by stevehipwell** [kubernetes-sigs/karpenter#993]
### Description

**What problem are you trying to solve?**
In a CA backed node pool implementation a minimum capacity can be reserved, with CA automatically scaling once those resources are no longer sufficient. This can also be augmented with preemptible pods to support there being free capacity available for new pods. Finally CA limits the total amount of resources used by the cluster to constrain all of this. This isn't perfect as there is no way to configure an upper limit on preemptible resources which is lower than the cluster maximum enforced by CA.

In the Karpenter context the CA implementation doesn't directly map as Karpenter node pools are heterogenous, instead of homogenous like CA needs, resulting in `n` nodes not mapping to a concrete amount of resources (a naive implementation would just reserve the smallest large instance with the required resources). Additionally given the fact that a single karpenter node pool can replace many separate legacy node pools there isn't the same topological distinction between pools required to make this work. But Karpenter does have the architecture to solve this problem in a better and more flexible way than CA can.

My proposal is to add a new CRD along the lines of `PreemtiblePod` which would have a resource field, selectors compatible with a pod and a field to set the mode. The default behaviour would be act like a standard preemptible pod and require Karpenter keep enough nodes ready for the equivalent pods to be scheduled immediately, this behaviour could be modified to act like the minimum node pool size as supported by CA. It would then be possible to add support to ignore these resources due to specific constraints such as a CRON pattern or provisioned resources etc. As these resources would be KRM they could both be structural to the K8s cluster or dynamically added as and when required.

**How important is this feature to you?**
Today this would be a nice to have as the CA minimum nodes functionality is heavily used by our cluster operators. Moving forwards the more advanced patterns would be potentially game changing in the context of batch workloads where we need to make sure we've got resources available before running a job.

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

**Comment by gmoirod** [comment]
A good use case for that would be CI jobs.
Often we see CI runner pods in K8S spinned up for jobs.
And CI jobs have to start quickly. We cannot afford to wait for a new node to be ready.
So we need reserved capacity dedicated for this.

**Comment by stevehipwell** [comment]
@morgayle-ngt I'm not sure that this is fully covered by the discussion in #749. This request is for Karpenter to support a baseline compute capacity and also a mechanism like the CA `provisioningrequests.autoscaling.x-k8s.io` CRD to temporarily request resources to be available. The important thing here is that Karpenter works differently to other systems so when reserving compute it should be modeled in a complimentary way.

In the Karpenter context a baseline compute capacity would need to be modeled on a pod spec to make sure that resource units and topology zones (likely cloud AZs, but potentially hosts) could be modeled; e.g. I might want my cluster to alway have 16 cores in blocks no smaller than 8 cpus per AZ even if there wasn't the workload to use them.

For the temporary resource requests, the same requirements are relevant, but there is also the operational aspect where such a request wants a TTL both to fit nicely with IaC patterns but also to limit the blast radius when something goes wrong making use of the compute.

FYI all of this depends on #745.

I'm also not sure what you mean by a "dual NodePool strategy" or its relevance to this, do you have docs?

**Comment by jonathan-innis** [comment]
I don't want to get this confused with the [reserved capacity feature]() that was just implemented and released so I'm going to suggest that we retitle this to something like "Pod Overhead"

/retitle Add support for Pod Overhead as an alternative to negative priority pods
